frontend/test/resolution/testManage.cpp

```diff
@@ -0,0 +1,265 @@
+  auto ctx = buildStdContext();
+  ErrorGuard guard(ctx);
+  std::string program =
+    R"""(
+    var global: int;
+    record R : contextManager {
+      proc type contextReturnType type do return int;
+
+      proc enterContext() ref {
+        return global;
+      }
+      proc exitContext(in error: owned Error?) {}
+    }
+    manage new R() as ref res {
+
+    }
+    )""";
+
+  auto x = resolveTypesOfVariables(ctx, program, {"res"}).at("res");
+  assert(!x.isUnknownOrErroneous());
+  assert(x.type()->isIntType());
+  assert(x.isRef());
+}
+
+static void testResourceByConstRef() {
+  // test saving the resource into a variable.
+  auto ctx = buildStdContext();
+  ErrorGuard guard(ctx);
+  std::string program =
+    R"""(
+    var global: int;
+    record R : contextManager {
+      proc type contextReturnType type do return int;
+
+      proc enterContext() ref {
+        return global;
+      }
+      proc exitContext(in error: owned Error?) {}
+    }
+    manage new R() as const ref res {
+
+    }
+    )""";
+
+  auto x = resolveTypesOfVariables(ctx, program, {"res"}).at("res");
+  assert(!x.isUnknownOrErroneous());
+  assert(x.type()->isIntType());
+  assert(x.isConst() && x.isRef());
+}
+
+static void testInferReturn() {
+  // test saving the resource into a variable.
+  auto ctx = buildStdContext();
+  ErrorGuard guard(ctx);
+  std::string program =
+    R"""(
+    record R : contextManager {
+      proc enterContext() {
+        return 42;
+      }
+      proc exitContext(in error: owned Error?) {}
+    }
+    manage new R() as res {}
+    )""";
+
+  auto x = resolveTypesOfVariables(ctx, program, {"res"}).at("res");
+  assert(x.type()->isIntType());
+}
+
+static void testReturnIntentOverload() {
+  // test saving the resource into a variable.
+  auto ctx = buildStdContext();
+  ErrorGuard guard(ctx);
+  std::string program =
+    R"""(
+    module M {
+      var global: int;
+      record R : contextManager {
+        proc type contextReturnType type do return int;
+
+        proc enterContext() {
+          return global;
+        }
+        proc enterContext() ref {
+          return global;
+        }
+        proc enterContext() const ref {
+          return global;
+        }
+        proc exitContext(in error: owned Error?) {}
+      }
+      manage new R() as byVal, new R() as ref byRef, new R() as const ref byConstRef {
+
+      }
+    }
+    )""";
+
+  setFileText(ctx, "test.chpl", program);
+  auto& modules = parsing::parse(ctx, UniqueString::get(ctx, "test.chpl"), UniqueString());
+  assert(modules.size() == 1);
+  auto& rr = resolveModule(ctx, modules[0]->id());
+
+  bool found;
+
+  auto byVal = findVariable(modules[0], "byVal");
+  auto byValParent = parsing::parentAst(ctx, byVal)->toAs()->symbol();
+  assert(byValParent);
+  assert(rr.hasAst(byValParent));
+  found = false;
+  for (auto aa : rr.byAst(byValParent).associatedActions()) {
+    if (aa.action() == AssociatedAction::ENTER_CONTEXT) {
```

Comment:
Should check the `EXIT_CONTEXT` actions as well. Doesn't necessarily have to be in this test since the point of this one is return intent overloading, but it could be.